Add vector search as new operation type to search

VijayanB opened the pull request:
### Description
For Vector search, we have to compare actual response's hits with expected neighbors to estimate recall for search. However, there is no provision to return recall similar to throughput. Hence, we will publish recall metrics like recall@k, recall@r as part of meta object. We also time those recall calculation and publish in ms to meta object.

Added tests to verify results are as expected.

### Issues Resolved
Part of  https://github.com/opensearch-project/opensearch-benchmark/issues/103 

### Testing
- [x] New functionality includes testing

[Describe how this change was tested]
```
make test 

.venv/lib/python3.8/site-packages/pkg_resources/__init__.py:2868
.venv/lib/python3.8/site-packages/pkg_resources/__init__.py:2868
  /Users/balasvij/PycharmProjects/knn/opensearch-benchmark/.venv/lib/python3.8/site-packages/pkg_resources/__init__.py:2868: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================= 1191 passed, 5 skipped, 3 warnings in 16.93s =================


---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

navneet1v commented on osbenchmark/worker_coordinator/runner.py:
```diff
@@ -967,6 +968,75 @@ async def _scroll_query(opensearch, params):
+                """
+                correct = 0.0
+                if neighbors is None:
+                    return 0.0
```
lets add a log message stating that neighbors are not found.

VijayanB replied:
Ack

navneet1v commented on osbenchmark/worker_coordinator/runner.py:
```diff
@@ -967,6 +968,75 @@ async def _scroll_query(opensearch, params):
+                return correct / total_num_of_results
+
+            doc_type = params.get("type")
+            response = await self._raw_search(opensearch, doc_type, index, body, request_params, headers=headers)
+            recall_processing_start = time.perf_counter()
+            response_json = json.loads(response.getvalue())
+            if detailed_results:
+                props = parse(response, ["hits.total", "hits.total.value", "hits.total.relation", "timed_out", "took"])
+                hits_total = props.get("hits.total.value", props.get("hits.total", 0))
+                hits_relation = props.get("hits.total.relation", "eq")
+                timed_out = props.get("timed_out", False)
+                took = props.get("took", 0)
+
+                result.update({
+                    "hits": hits_total,
+                    "hits_relation": hits_relation,
+                    "timed_out": timed_out,
+                    "took": took
+                })
```
what will happen here if hits are empty? and search is giving 200.

VijayanB replied:
Updated.

VijayanB replied:
Added test case where hits contain empty list.

jmazanec15 commented on osbenchmark/worker_coordinator/runner.py:
```diff
@@ -967,6 +968,76 @@ async def _scroll_query(opensearch, params):
+                    results: list containing ids of results returned by OpenSearch.
+                    neighbors: set containing ids of the actual neighbors
+                    neighbors for a set of queries
+                    r: number of top results to check from the neighbors
```
Can you add more details on this here? Will give future devs more guidance

VijayanB replied:
Ack

VijayanB commented:
@jmazanec15 @navneet1v I removed rcall@r, and, kept only recall@k, where, we will compare predictions with top k neighbors from neighbor's list.

navneet1v commented:
> @jmazanec15 @navneet1v I removed rcall@r, and, kept only recall@k, where, we will compare predictions with top k neighbors from neighbor's list.

can we have recall@1 always getting calculated?

navneet1v commented on osbenchmark/worker_coordinator/runner.py:
```diff
@@ -967,6 +968,90 @@ async def _scroll_query(opensearch, params):
+                min_num_of_results = min(top_k, len(neighbors))
+                truth_set = neighbors[:min_num_of_results]
+                for j in range(min_num_of_results):
+                    if j >= len(predictions):
+                        break
```
lets add a log here too, which specify what happened.

VijayanB replied:
Ack

VijayanB commented:
> > @jmazanec15 @navneet1v I removed rcall@r, and, kept only recall@k, where, we will compare predictions with top k neighbors from neighbor's list.
> 
> can we have recall@1 always getting calculated?

Added recall@1 and updated testcase.

IanHoang commented on osbenchmark/worker_coordinator/runner.py:
```diff
@@ -967,6 +968,97 @@ async def _scroll_query(opensearch, params):
+                Args:
+                    predictions: list containing ids of results returned by OpenSearch.
+                    neighbors: list containing ids of the actual neighbors
+                    neighbors for a set of queries
```
Nit: Should this just be `list containing ids of the actual neighbors for a set of queries` is this part of the sentence above?

VijayanB replied:
Ack

IanHoang commented on osbenchmark/worker_coordinator/runner.py:
```diff
@@ -967,6 +968,97 @@ async def _scroll_query(opensearch, params):
+                truth_set = neighbors[:min_num_of_results]
+                for j in range(min_num_of_results):
+                    if j >= len(predictions):
+                        self.logger.info("No more neighbors in prediction to compare against ground truth.\n"
```
Pylint raises a warning about this line since you're using fstrings in logging statements. Please conform and use lazy % interpolation. 
```
self.logger.info("No more neighbors in prediction to compare against ground truth.\n"
    "Total neighbors in prediction: %s. \n"
    "Total neighbors in ground truth: %s", len(predictions), min_num_of_results)
```

VijayanB replied:
Ack

IanHoang reviewed (changes requested):
Left a couple comments
